docs: update developer resources (#286)
urbanopt/geojson-modelica-translator: docs/developer_resources.rst

Design Overview
===============

The GMT is designed to create an arbitrary number of user-configured models connected to other user-configured models to represent a district energy system.

> GMT has "building blocks" that it uses to define and connect models, which currently include: ETSes, Loads, Networks, and Plants.

Reviewer: ```suggestion
GMT has "building blocks" that it uses to define and connect models, which currently include: Energy Transfer Stations (ETSs), Loads, Networks, and Plants.
```

> 

Reviewer 2: Thanks for the puzzle image.

I think it would be more clearer to modify titles with:

- Thermal loads block instead of load
- ETS block instead of ETS 
- District distribution network block instead of Network
- Central thermal plants block instead of plants

> .. image:: images/models-overview.png

Reviewer: noice image!

> Each block type is a collection of models, e.g. the Loads includes time series and spawn models. A model in GMT refers to an abstracted modelica model. It generates modelica code (the modelica model) and is used to define connections to other models.

Reviewer: ```suggestion
Each block type is a collection of models, e.g. the Loads includes time series and spawn models. A model in GMT refers to an abstracted Modelica model. It generates Modelica code (the Modelica model) and is used to define connections to other models.
```

> Each block type can "connect" (note this connect does not refer to modelica's connect) to other specific block types. For example, you can connect a load to an ets, but you cannot connect a load to a network.

Reviewer: ```suggestion
Each block type can "connect" (note this "connect" does not refer to Modelica's concept of "connect") to other specific block types. For example, you can connect a load to an ETS, but you cannot connect a load to a network.
```

Each block type has a corresponding directory inside of :code:`geojson_modelica_translator/model_connectors`, which contains its different model types (e.g. for Loads it contains the Time Series model and others).

Because models are different, even within a block type (e.g. different properties and maybe even ports), the GMT uses the concept of couplings for connecting models. Couplings define how two *specific* models connect in modelica.
For example, a coupling could define how the time series load actually connects to the heating indirect ETS.

    -- As an aside, if the GMT reached a point where all models within a block type implemented the same interface then couplings would not be necessary.

Adding New Models
=================

> To add a new model you have to do a few things

Reviewer: ```suggestion
To add a new model you have to do the following:
```

1. Define the model's python class: First, create a new python file and class under its respective directory in model_connectors. Follow the patterns of existing classes.

> 2. Create coupling files: For every model that it should be able to be linked to, create a <ModelA>_<ModelB> directory in the couplings directory. The two files ComponentDefinitions.mopt and ConnectStatements.mopt must exist in this directory.

Reviewer: ```suggestion
2. Create coupling files: For every model that can be linked to, create a <ModelA>_<ModelB> directory in the couplings directory. The two files ComponentDefinitions.mopt and ConnectStatements.mopt must exist in this directory. See more information on the content of the coupling files below in the *Couplings* sections.
```

Couplings
=========

> A coupling defines the modelica code necessary for interfacing two specific models, e.g. a time series load and heating indirect ETS.

Reviewer: ```suggestion
A coupling defines the Modelica code necessary for interfacing two specific models, e.g. a time series load and heating indirect ETS.
```

Each coupling is unique in its requirements:

- what additional components are necessary, for example there might be some sensor between system A and B, or maybe B requires a pump when A is a specific model type
- what ports are connected, for example connecting ports of model A and model B

Thus each coupling must define two template files, ComponentDefinitions.mopt and ConnectStatements.mopt, respectively. These files must be placed in the directory :code:`couplings/templates/<model A>_<model B>/`.

> In general, the order of the names should follow the order of system types if you laid out the district system starting with loads on the far left and plants on the far right (e.g. load before ets, ets before network, network before plant)

Reviewer: ```suggestion
In general, the order of the names should follow the order of system types if you laid out the district system starting with loads on the far left and plants on the far right (e.g. load before ETS, ETS before network, network before plant)
```

> In general, the order of the names should follow the order of system types if you laid out the district system starting with loads on the far left and plants on the far right (e.g. load before ets, ets before network, network before plant)

Reviewer 2: I don't seem to fully understand the line starting from In general till before plant.

Does the user need to order the component's names in order to successfully couple it?

Author: Sorry I missed this comment earlier. The user of the GMT does not need to worry about orderings, but the developer does. Specifically, the developer needs to know the ordering when making new `couplings/templates/<model A>_<model B>/` directories. I will update my notes on this section to try to make it a bit more clear

When rendering the district system model file, it must:

> 1. call to_modelica() for each model to generate its modelica code

Reviewer: ```suggestion
1. call to_modelica() for each model to generate its Modelica code
```

> 2. render the coupling partial templates (ie the modelica code required for couplings)

Reviewer: ```suggestion
2. render the coupling partial templates (ie the Modelica code required for couplings)
```

> 3. render the model instance definition (ie the modelica code which instantiates the model)

Reviewer: ```suggestion
3. render the model instance definition (ie the Modelica code which instantiates the model)
```

> 4. insert the coupling partials and model instance definitions into the district modelica file

Reviewer: ```suggestion
4. insert the coupling partials and model instance definitions into the district Modelica file
```

> Refer to DistrictEnergySystem.mot and district.py for reference.

Reviewer 2: is there a possibility to include the path a "hyperlink" for the template at the GMT?

Summary of Templating Contexts
++++++++++++++++++++++++++++++

Model Definition
**********

> Each model generates one or more modelica files to define its model. The templating context is implementation dependent, so refer to its :code:`to_modelica()` method.

Reviewer: ```suggestion
Each model generates one or more Modelica files to define its model. The templating context is implementation dependent, so refer to its :code:`to_modelica()` method.
```

Component Definitions
*********************

This is the template which defines new components/variables necessary for a coupling. It has access to:

- :code:`globals`: global variables (those defined in the district.py, such as medium_w = MediumW)

> - :code:`coupling`: contains the coupling id, as well as references to the coupled models under their respective types (e.g. coupling.load.id or coupling.network.id). You should append :code:`coupling.id` to any variable identifiers to prevent name collisions. For example, instead of just writing :code:`parameter Modelica.SIunits.MassFlowRate mDis_flow_nominal` you should do :code:`parameter Modelica.SIunits.MassFlowRate mDis_flow_nominal_{{ coupling.id }}` as well as any place where you would reference that variable.

Reviewer 2: can we change the components naming mechanism? because the larger the district energy model is" more buildings/ETS are connected "  the harder to debug it with the "alphabetics-numbers" mixed names.

Author: I have created #288 to track this request 👍
